### PR TITLE
Topic/reflection without remorse

### PR DESCRIPTION
**Pull request by ethul**
I've been kicking around these changes for a while, but I wanted to get them together in this PR to spark a discussion and get a feel for what others think about taking a new direction with the free monad implementation.

The main change is that the implementation is in the spirit of the implementation described in the paper [Reflection without Remorse](http://okmij.org/ftp/Haskell/zseq.pdf). I think going this route is beneficial because it alleviates quadratic complexity issues in both left-associated binding and in observability. [Mandubian's blog](http://mandubian.com/2015/04/09/freer) gives some additional of these issues in terms of Scala.

Attached is a preliminary benchmark for left-associated binding of the current free monad implementation at v0.6.0, and then the new implementation based on Reflection without Remorse.

left-associated-binds

Even though these results are preliminary, I think it is clear on what gains are possible with this implementation. On that note, I am totally open to discussion on the underlying data structure used to store up the binds of the free monad. I currently have a linked-based structure, which seems to perform well. I tried `Data.Sequence`, but I couldn't get the performance. In any case, I am up for a discussion on this.

Also, the free monad is defined more [operationally](https://github.com/atzeus/reflection-without-remorse/blob/8ea7ad4c58230d126730fa005d996184e58eb00a/Control/Monad/Operational/Reflectable.hs), which allowed for a simplified API. We no longer really need `FreeC` since most of the `Free` functions no longer require a functor instance. I think this makes the free monad more straightforward to use and removes duplicating functions implemented for both `Free` and `FreeC`.

I know that these are significant changes to the free monad internals, as well as some breaking changes. So I am up for any feedback, comments, questions, and/or suggestions.

### CODE REVIEW

**Comment by garyb** [comment]
Wow, this looks really interesting! I can't really see any downsides in what you're proposing in that it gives us better performance and a simpler API, but I primarily understand `Free` in how its used, not it's "deeper" meaning, so not sure if there are any cons on that side of things.

My only (rather silly) comment after having a quick look is can we keep the type synonym as just `Natural`? I've been using it a lot lately, and `NaturalTransformation` seems a bit wordy now :smile:

**Comment by paf31** [comment]
This is exciting :)

File: src/Control/Monad/Free.purs [hunk 1]
```diff
@@ -1,125 +1,79 @@
...
-  wrap = Free
+instance freeMonadTrans :: MonadTrans Free where
+  lift = liftF
 
--- | Lift an action described by the generating functor `f` into the monad `m`
--- | (usually `Free f`).
-liftF :: forall f m a. (Functor f, Monad m, MonadFree f m) => f a -> m a
-liftF = wrap <<< map pure
+instance freeMonadRec :: MonadRec (Free f) where
+  tailRecM k a = k a >>= either (tailRecM k) pure
```
Comment (paf31) [inline src/Control/Monad/Free.purs]: Don't we need a function literal here to stop the recursive call being made?

Comment (paf31) [inline src/Control/Monad/Free.purs]: Oh wait, partial application, never mind.

File: src/Control/Monad/Free.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  this.value = value;
+  this.next = null;
+}
+
+function append(list1) {
+  return function(list2){
+    if (list1.size === 0) return list2;
+    else if (list2.size === 0) return list1;
+    else {
+      list1.last.next = list2.head;
```
Comment (paf31) [inline src/Control/Monad/Free.js]: Could you please add a comment explaining why the mutation is safe?

Comment (paf31) [inline src/Control/Monad/Free.js]: Is it worth it/possible to use `ST` to make it clear?

**Comment by jdegoes** [comment]
I'm a big fan of this approach. A few comments:

* Obviously this approach requires some unsafety in the implementation, which is totally OK as long as we can hide that in the external API.
* I'm a fan of minimizing unnecessary changes, e.g. `Natural` versus `NaturalTransformation`, and also retaining some helper functions that this PR doesn't have (e.g. `bindF`). Obviously, we'd want to delete the dual versions of functions, but beyond that, I don't think the external API has to change that much.

File: src/Control/Monad/Free.purs [hunk 3]
```diff
@@ -1,125 +1,79 @@
...
 
-instance applyFree :: (Functor f) => Apply (Free f) where
-  apply = ap
+instance freeFunctor :: Functor (Free f) where
+  map k f = f >>= return <<< k
 
-instance applicativeFree :: (Functor f) => Applicative (Free f) where
-  pure = Pure
+instance freeBind :: Bind (Free f) where
+  bind (Free v s) k = Free v (snoc s (ExpF (unsafeCoerceBind k)))
```
Comment (paf31) [inline src/Control/Monad/Free.purs]: This call to `snoc` can't be safe, can it? It modifies `s` in place.

**Comment by paf31** [comment]
If you need a structure with fast `uncons`, `snoc` and `append`, maybe a difference list would work? Or maybe a pair of lazy lists where you move elements across as needed?

File: src/Control/Monad/Free.purs (hunk 3 above)
Comment (jdegoes) [inline src/Control/Monad/Free.purs]: You can take the same Free and bind it once, and also bind it twice, and throw away the latter one. Now the list size for the first one will be wrong. So it's not safe.

Comment (paf31) [inline src/Control/Monad/Free.purs]: Sorry to be clear, we're saying "externally unsafe" here, right? :smile:

**Comment by jdegoes** [comment]
One more: I would recommend using a PureScript data structure rather than a native one. Performance penalty should be quite small and the additional type safety and readability would be nice. A difference list (which does not exist yet in PureScript?) does seem like a possible first choice.

**Comment by paf31** [comment]
https://github.com/Thimoteus/purescript-difflists

:smile:

**Comment by jdegoes** [comment]
> https://github.com/Thimoteus/purescript-difflists

I had no idea. :smile: I would not recommend that representation in PureScript (I don't think it will have good performance characteristics), instead I would recommend `data DList a = DList (List a) (List a)`. Also the project has no license file.

**Comment by ethul** [comment]
Thank you everyone for taking a look at these changes. I am excited by the direction this is going.

In terms of the API changes, we can definitely keep `Natural` over `NaturalTransformation`. Should we pull `Natural` out into its own module or leave it in `Data.Coyoneda`? Or leave `NaturalTransformation` and define an alias in `Data.Coyoneda`? I am open to any option we decide on.

Regarding `bindF`, I wrote an implementation for that but I ended up just writing it in terms of `foldMapF`

```purescript
bindF :: forall f g a. Free f a -> NaturalTransformation f (Free g) -> Free g a
bindF fa t = foldMapF t fa
```

Does `foldMapF` fit the use-case for `bindF`? If so, I am unsure if we need `bindF`.

Regarding the underlying data structure, I am definitely open to changing this. I originally tried purescript's `Data.Sequence`, but it did not end up with having good performance here. Below is an updated chart with those numbers.

left-associated-binds--data-sequence

I didn't dig very deep on why this wasn't performing well. However, I'd be open to taking a second look. Also, using a difference list sounds appealing. I will test out a few options and see what works best. Any other thoughts or options are most welcome.

**Comment by paf31** [comment]
I think `bindF` as `foldMap` makes sense. `foldMap` is expressing the adjointness of the free monad construction to the "only keep the functor" operation, so it makes more sense to have that than `bindF` I think. The `free` library calls it `foldFree`.

**Comment by puffnfresh** [comment]
This PR is amazing :rainbow:

**Comment by ethul** [comment]
:)

@paf31 Sounds good about `bindF`. And I am open to renaming `foldMapF` to `foldFree` or maybe even `foldF`. But I am just as happy to leave it as-is.

**Comment by Thimoteus** [comment]
@paf31 @jdegoes I wasn't expecting anyone to use my difflist package, but I'm updating it now (currently adding benchmarks, already added an MIT license that'll come in the next push). And my idea for it was to support difference lists for a variety of "cons"able structures.

**Comment by jdegoes** [comment]
> And my idea for it was to support difference lists for a variety of "cons"able structures.

I may be wrong, but I think a singly-linked list (aka `List`) is possibly the only structure which can benefit from this technique, so I'd suggest specializing specifically to `List`.

> EDIT: I updated it. Also @jdegoes are you thinking of the first type of difference list as explained here?

No, it's the same idea, just represented differently. `data DList a = DList (List a) (List a)` has two lists, one which represents prepends and one which represents appends. You get all the nice properties of a difference list without building up the very complex structure of the more classic representation of `DList`.

**Comment by ethul** [comment]
I've been working on a bare-bones implementation of a `DList` as well as a catenable list built on top of the `DList` in the repo [ethul/purescript-catenable](https://github.com/ethul/purescript-catenable). @jdegoes I am not sure if this was more or less the implementation you had in mind.

However, the [linkAll](https://github.com/ethul/purescript-catenable/blob/master/src/Data/CList.purs#L36-L41) function is not currently stack safe. I am wondering if we can come up with some ideas of how to make this tail recursive. Any thoughts would be greatly appreciated.

Also, it is possible to get `linkAll` to `O(1)` amortized time if we introduce laziness. I haven't gotten that far yet, but it might be a good next step. Or maybe there is a better way to do this entirely.

**Comment by ethul** [comment]
@paf31 I've been working on a stack safe version of [linkAll](https://github.com/ethul/purescript-catenable/blob/8910199300be9ce03048cfcc4315503fbca08ddc/src/Data/CList.purs#L52-L69) using the `purescript-safely` module. However, it still seems to be giving me a stack overflow. Is there something you can spot that is amiss with how I've got it wired up? I appreciate any thoughts or pointers.

**Comment by paf31** [comment]
@ethul That's still pretty experimental, but it only attempts to make the binds safe, not the rest of the computation. I've seen it fail before for things like `traverse`, where you build up a big data structure in the result. Maybe that's the issue? Can you see where the stack overflow is occurring?

**Comment by ethul** [comment]
@paf31 Thanks for taking a look. Noted on the experimental status. It is possible that the issue is where the data is built up. I will have to dig more into it. However, the stack overflow trace produced is:

```bash
RangeError: Maximum call stack size exceeded
    at new Bind (/path/purescript/purescript-free/output/Prelude/index.js)
    at /path/purescript/purescript-free/output/Control.Coroutine/index.js:164:16
    at null.__superclass_Prelude.Bind_1 (/path/purescript/purescript-free/output/Control.Coroutine/index.js:158:44)
    at /path/purescript/purescript-free/output/Data.CList/index.js:89:88
    at /path/purescript/purescript-free/output/Data.CList/index.js:89:120
    at /path/purescript/purescript-free/output/Data.CList/index.js:89:120
    at /path/purescript/purescript-free/output/Data.CList/index.js:89:120
    at /path/purescript/purescript-free/output/Data.CList/index.js:89:120
    at /path/purescript/purescript-free/output/Data.CList/index.js:89:120
    at /path/purescript/purescript-free/output/Data.CList/index.js:89:120
```

So it looks to fail at the `return new Prelude.Bind(function () {` in

```javascript
var bindCo = function (__dict_Functor_28) {
    return function (__dict_Monad_29) {
        return new Prelude.Bind(function () {
            return applyCo(__dict_Functor_28)(__dict_Monad_29);
        }, function (_43) {
            return function (f) {
                if (_43 instanceof Bind) {
                    return Data_Exists.runExists(function (_23) {
                        return bound(_23.value0)(function (x) {
                            return bound(function (_22) {
                                return _23.value1(x);
                            })(f);
                        });
                    })(_43.value0);
                };
                return bound(function (_24) {
                    return _43;
                })(f);
            };
        });
    };
};
```

**Comment by paf31** [comment]
But that's just the top stack frame right? Looks like it gets stuck at `/purescript/purescript-free/output/Data.CList/index.js:89:120`

**Comment by ethul** [comment]
@paf31 You're right. That is the line `return Prelude[">>="]...` in

```javascript
var foldrM = function (__dict_Monad_1) {
    return function (k) {
        return function (b) {
            return function (dlist) {
                var _94 = Data_DList.uncons(dlist);
                if (_94 instanceof Data_Maybe.Nothing) {
                    return Prelude.pure(__dict_Monad_1["__superclass_Prelude.Applicative_0"]())(CNil.value);
                };
                if (_94 instanceof Data_Maybe.Just) {
                    return Prelude[">>="](__dict_Monad_1["__superclass_Prelude.Bind_1"]())(foldrM(__dict_Monad_1)(k)(b)(_94.value0.value1))(k(_94.value0.value0));
                };
                throw new Error("Failed pattern match: " + [ _94.constructor.name ]);
            };
        };
    };
};
```

**Comment by ethul** [comment]
@paf31 Good catch. So the trick is to make that `foldrM` call tail recursive I think.

**Comment by ethul** [comment]
Sorry for the delay here.

I've switch over to a catenable list implementation for the underlying data structure. I tested with a [strict version](https://github.com/ethul/purescript-catenable/blob/cbd240d7d6e38045d2fb9a4c4f6d213d46735569/src/Data/CList.purs) and [lazy version](https://github.com/ethul/purescript-catenable/blob/e1840aa78cb7c258895ac8cce8e3edfe8adbe1f5/src/Data/CList.purs).

left-binds

Adding laziness was suppose to ideally improve the cost of the `linkAll` function, but it looks like the overhead is not worth it (given my preliminary benchmarks).

Most of this is coming from [Purely Functional Data Structures](http://www.cs.cmu.edu/~rwh/theses/okasaki.pdf), section 7.2.1.

What does everyone think? I am inclined to go with the strict version. Seems pretty good in terms of performance. And comparable to the `new` data structure in the figure. It is more costly in terms of memory than `new` (I couldn't get it to push 6 million left-binds). In any case, if this works for everyone I will revert to the strict `CList` and make the other changes as mentioned above:

 - Add `Natural` back to `Data.Coyoneda` (perhaps as an alias for `NaturalTransformation`)
 - Rename `foldMap` to `foldFree`

Also, any suggestions on names for what I am calling `CList` and `DList`? I was thinking potentially `CatList` or `CatenableList` for `CList`. But I am open to ideas for these.

**Comment by garyb** [comment]
Looks great to me at least :+1: 

Re: `Natural`, having it as `Natural` in `Data.Natural` or `Data.NaturalTransformation` is probably my preference.

**Comment by paf31** [comment]
Strict version sounds good to me :+1: 

> I couldn't get it to push 6 million left-binds

How does it fare if you use something like `forever`? Does it run out of memory, or do things get cleaned up incrementally? (Sorry, I still need to read the paper...)

**Comment by ethul** [comment]
I've updated the PR to include the strict `CatList` implementation from [purescript-catenable-lists](https://github.com/ethul/purescript-catenable-lists), and made the updates mentioned above.

Updated benchmarks:
left-binds

@garyb I took your suggestion to keep `Natural` in `Data.NaturalTransformation`, adding it as an alias.

@paf31 In running the tests on a large inputs I have to run node with `--max_old_space_size=<my computer's maximum about of memory>`. Otherwise node segfaults with an out of memory error. So it seems things are not getting cleaned up. I haven't tried `forever` yet, but image it will yield the out of memory error. I can look into this further though to see if there is a way to avoid this.

Also, I'd like to add a few more benchmarks now that we have the `CatList` in place. I've [benchmarked](https://github.com/ethul/purescript-catenable-lists#benchmarks) the operations on `CatList` separately however using @hdgarrood's great `purescript-benchotron` package. I'd like to add such benchmarks here too.

Let me know if there are any other additions and/or changes.

Thanks again for taking a look at this.

**Comment by hdgarrood** [comment]
Ahhh, thank you! :) Just out of interest, how are you generating those graphs? They look much nicer than the ones I managed to generate!

**Comment by ethul** [comment]
@hdgarrood They are nothing too fancy, just produced from a spreadsheet using [libreoffice](http://www.libreoffice.org/)

**Comment by hdgarrood** [comment]
Ah, interesting. Thanks :)

**Comment by ethul** [comment]
@hdgarrood welcome

**Comment by paf31** [comment]
I think we'd definitely want to be able to support `forever` without issues. Again, I need to read the paper, but I'd hope the data structure would hover around the same size.

**Comment by paf31** [comment]
`CatList` looks fantastic by the way :+1:

**Comment by ethul** [comment]
@paf31 Thanks! :)

Regarding `forever`, even with the `Free` implementation using `Gosub` (v0.6.0), node would segfault with an out of memory error. I believe I was able to get to just over 3 million left-binds. So using `CatList` we can double that, but eventually node will run out of memory.

But let me give `forever` a try to see what happens. Actually, what is the precise test case you have in mind? Is it to build up left-binds using forever? Or something like:

```
forever $
  make a program
  interpret the program
```

**Comment by paf31** [comment]
No, just something like

```purescript
loop :: forall a. Free TeletypeF a
loop = forever $ writeLine "blah"
```

using the `Teletype` example. Then, building and interpreting that should both work, ideally. I'm pretty sure they do now.

**Comment by ethul** [comment]
@paf31 Gotcha. Running [the program](https://gist.github.com/ethul/ee9a7ef6238903df4579) now

**Comment by paf31** [comment]
Oh, one other thing - would you consider moving `catenable-lists` to the core libraries? I think it would be a great addition in its own right, but generally things in core only depend on other things in core, so that we can be sure they are all in sync and build against the latest compiler.

If you'd rather not, we can figure something else out.

**Comment by ethul** [comment]
@paf31 I'd be happy to move it over to core. And it looks like bower has an unregister now. So hopefully it would be an easy endpoint switch :)

**Comment by ethul** [comment]
@paf31 Okay, I've unregistered it from bower. Let me know if I can transfer over the repo. I will reregister it after the transfer.

**Comment by paf31** [comment]
@ethul Thanks, you should get an invite.

**Comment by ethul** [comment]
@paf31 Transfer done and registered to bower.

**Comment by paf31** [comment]
:+1: Thanks!

**Comment by ethul** [comment]
Welcome!

**Comment by paf31** [comment]
I've added you as an admin on the `free` and `catenable-lists` repos too.

**Comment by ethul** [comment]
@paf31 Awesome. Thanks!

**Comment by ethul** [comment]
@paf31 By the way, `Forever` has been running now for a good amount of time and memory usage has been stable.

**Comment by paf31** [comment]
:+1: Excellent

File: src/Control/Monad/Free.purs [hunk 4]
```diff
@@ -132,52 +82,63 @@ bindFC fa t = bindF fa (liftCoyonedaTF t)
...
+                        Just (Tuple h t) -> toView (unsafeCoerceFree (concatF ((runExpF h) a) t))
+       Bind f k -> Bind f (\a -> unsafeCoerceFree (concatF (k a) s))
+  where
+  concatF :: Free f Val -> CatList (ExpF f) -> Free f Val
+  concatF (Free v l) r = Free v (l <> r)
+
+  runExpF :: forall f. ExpF f -> (Val -> Free f Val)
+  runExpF (ExpF k) = k
+
+unsafeCoerceVal :: forall a. Val -> a
```
Comment (paf31) [inline src/Control/Monad/Free.purs]: Does it make sense to move these into `where` blocks, or are they reused a lot?

Comment (ethul) [inline src/Control/Monad/Free.purs]: I'd be up to move all the unsafe coerce functions into appropriate where blocks. I think it would make the code a bit cleaner.

**Comment by garyb** [comment]
Is there anything else that needs doing here? It'd be great to get this in before the new Halogen, as it would simplify a lot of the variations I have in there for `Coyoneda`.

**Comment by ethul** [comment]
@garyb I was thinking of add some more benchmarks with `purescript-benchotron`, but I can bring these in in a separate PR.

**Comment by ethul** [comment]
I've added the following benchmark images for reference:
left-associated-binds
right-associated-binds

**Comment by garyb** [comment]
Awesome!

Something I meant to ask before, what's the performance like for much smaller numbers of binds? Is there much/any overhead vs the old method? A bit harder to measure I guess due to much greater variability.

**Comment by ethul** [comment]
@garyb Here are the benchmarks at a smaller scale:
left-associated-binds--small
right-associated-binds--small

**Comment by ethul** [comment]
@garyb Note that I will remake these benchmarks with variance once I switch it over to using `purescript-benchotron`. But let me know if you'd like me to include that switch-over in this PR.

**Comment by garyb** [comment]
Perfect, thanks for taking the time to do those too. I don't mind whether we include the `benchotron` stuff or not in this PR :)

**Comment by ethul** [comment]
@garyb Welcome. Maybe I will save the `benchotron` changes for a separate PR then.

**Comment by garyb** [comment]
I'll merge this as it is so I can start using it in Halogen actually via hash-versioning (I have an addition I might propose via PR too). I'll hold off making a release until you're happy with the benchmark situation.

**Comment by garyb** [comment]
Thanks by the way, this is amazing! :tada:

**Comment by ethul** [comment]
Great! Thanks for merging. I will send a PR with the benchotron updates.
